File: backend/src/domain/orders/order.rb
require_relative 'order_item'
require_relative 'payment'
require_relative '../customers/address'

class Order
  attr_reader :id, :customer, :items, :payment, :address, :closed_at

  def initialize(id, customer, overrides = {})
    @id = id
    @customer = customer
    @items = []
    @address = overrides.fetch(:address) { Address.new(zipcode: '45678-979') }
  end

  def add_product(product)
    @items << OrderItem.new(order: self, product: product)
  end

  def total_amount
    @items.map(&:total).inject(:+)
  end

  def add_payment(payment_method)
    @payment = Payment.new(self, payment_method: payment_method)
  end

  def pay
    @payment.pay
    close
  end

  def invoice
    @payment.invoice
  end

  def to_s
    order_as_string = "Order #{id} - Total #{total_amount}\n"
    @items.each { |item| order_as_string << "#{item}\n" }
    order_as_string
  end

  private 

  def close(closed_at = Time.now)
    @closed_at = closed_at
    @items.each { |item| item.close() }
  end 
end

File: backend/src/Order.kt
package challenge

import java.util.*

/* rules
* Se o pagamento for para um item físico, você deverá gerar um shipping label para o mesmo ser colocado na caixa do envio;
*
Caso o pagamento seja uma assinatura de serviço, você precisa ativar a assinatura, e notificar o usuário através de e-mail sobre isto;

Caso o pagamento seja um livro comum, você precisa gerar o shipping label com uma notificação de que trata-se de um item isento de impostos
conforme disposto na Constituição Art. 150, VI, d.

Caso o pagamento seja de alguma mídia digital (música, vídeo), além de enviar a descrição da compra por e-mail ao comprador,
conceder um voucher de desconto de R$ 10 ao comprador associado ao pagamento.
* */

class Order(val customer: Customer, val address: Address) {

    private val items = mutableListOf<OrderItem>()

    var closedAt: Date? = null
        private set
    var payment: Payment? = null
        private set
    val totalAmount
        get() = items.sumByDouble { it.total }

    fun addProduct(product: Product, quantity: Int) {
        val productAlreadyAdded = items.any { it.product == product }
        if (productAlreadyAdded)
            throw Exception("The product have already been added. Change the amount if you want more.")

        items.add(OrderItem(product, quantity))
    }

    fun pay(method: PaymentMethod) {
        if (payment != null)
            throw Exception("The order has already been paid!")

        if (items.count() == 0)
            throw Exception("Empty order can not be paid!")

        payment = Payment(this, method)

        close()
    }

    private fun close() {
        closedAt = Date()
    }

    fun processOrderItems() {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }
}

data class OrderItem(val product: Product, val quantity: Int) {
    val total get() = product.price * quantity
}